app/src/main/java/org/wikipedia/language/LanguagesListViewModel.kt

```diff
@@ -0,0 +1,88 @@
+import org.wikipedia.util.log.L
+import java.util.*
+
+class LanguagesListViewModel : ViewModel() {
+
+    val siteListData = MutableLiveData<Resource<List<SiteMatrix.SiteInfo>>>()
+    private val suggestedLanguageCodes = WikipediaApp.getInstance().language().remainingAvailableLanguageCodes
+    private val nonSuggestedLanguageCodes = WikipediaApp.getInstance().language()
+        .appMruLanguageCodes.toMutableList().also {
+                it.removeAll(suggestedLanguageCodes)
+                it.removeAll(WikipediaApp.getInstance().language().appLanguageCodes)
+            }
+
+    private val handler = CoroutineExceptionHandler { _, throwable ->
+        L.e(throwable)
+    }
+
+    init {
+        fetchData()
+    }
+
+    private fun fetchData() {
+        viewModelScope.launch(handler) {
+            withContext(Dispatchers.IO) {
+                val siteMatrix = ServiceFactory.get(WikipediaApp.getInstance().wikiSite).getSiteMatrix()
+                val sites = SiteMatrix.getSites(siteMatrix)
+                siteListData.postValue(Resource.Success(sites))
+            }
+        }
+    }
+
+    fun getListBySearchTerm(context: Context, searchTerm: String?): List<LanguageListItem> {
+        val results = mutableListOf<LanguageListItem>()
+        val filter = StringUtils.stripAccents(searchTerm.orEmpty()).lowercase(Locale.getDefault())
+
+        addFilteredLanguageListItems(filter, suggestedLanguageCodes,
+                context.getString(R.string.languages_list_suggested_text), results)
+
+        addFilteredLanguageListItems(filter, nonSuggestedLanguageCodes,
+                context.getString(R.string.languages_list_all_text), results)
+
+        return results
+    }
+
+    private fun addFilteredLanguageListItems(filter: String, codes: List<String>, headerText: String,
+                                             results: MutableList<LanguageListItem>) {
+        var first = true
+        for (code in codes) {
+            val localizedName = StringUtils.stripAccents(WikipediaApp.getInstance().language().getAppLanguageLocalizedName(code).orEmpty())
+            val canonicalName = StringUtils.stripAccents(getCanonicalName(code).orEmpty())
+            if (filter.isEmpty() || code.contains(filter) ||
+                    localizedName.lowercase(Locale.getDefault()).contains(filter) ||
+                    canonicalName.lowercase(Locale.getDefault()).contains(filter)) {
+                if (first) {
+                    results.add(LanguageListItem(headerText, true))
+                    first = false
+                }
+                results.add(LanguageListItem(code))
+            }
+        }
+    }
+
+    fun getCanonicalName(code: String): String? {
+        var canonicalName = siteListData.value?.data?.find { it.code == code }?.localname
+        if (canonicalName.isNullOrEmpty()) {
+            canonicalName = WikipediaApp.getInstance().language().getAppLanguageCanonicalName(code)
+        }
+        return canonicalName
```

Comment:
This can be simplified to:
```
siteListData.value?.data?.find { it.code == code }?.localname.orEmpty().ifEmpty { WikipediaApp.getInstance().language().getAppLanguageCanonicalName(code) }
```

Reply:
yep